src/cargo/util/toml/embedded.rs

```diff
@@ -189,94 +189,112 @@ fn sanitize_name(name: &str) -> String {
 }
 
 #[derive(Debug)]
-struct Source<'s> {
+pub struct ScriptSource<'s> {
     shebang: Option<&'s str>,
     info: Option<&'s str>,
     frontmatter: Option<&'s str>,
     content: &'s str,
 }
 
-fn split_source(input: &str) -> CargoResult<Source<'_>> {
-    let mut source = Source {
-        shebang: None,
-        info: None,
-        frontmatter: None,
-        content: input,
-    };
+impl<'s> ScriptSource<'s> {
+    pub fn parse(input: &'s str) -> CargoResult<Self> {
+        let mut source = Self {
+            shebang: None,
+            info: None,
+            frontmatter: None,
+            content: input,
+        };
+
+        // See rust-lang/rust's compiler/rustc_lexer/src/lib.rs's `strip_shebang`
+        // Shebang must start with `#!` literally, without any preceding whitespace.
+        // For simplicity we consider any line starting with `#!` a shebang,
+        // regardless of restrictions put on shebangs by specific platforms.
+        if let Some(rest) = source.content.strip_prefix("#!") {
+            // Ok, this is a shebang but if the next non-whitespace token is `[`,
+            // then it may be valid Rust code, so consider it Rust code.
+            if rest.trim_start().starts_with('[') {
+                return Ok(source);
+            }
```

Comment:
Just a comment: This isn't quite the same behavior, as the compiler allows comments before the opening `[`. I realize parsing that would be difficult, and that particular scenario is a bit silly.

Reply:
This is existing code, just moved.  Ok if we put this in the tracking issue or an issue and track separately?

Reply:
Added this to the checklist: #12207